okio-resourcefilesystem/src/main/kotlin/okio/resourcefilesystem/ResourceFileSystem.kt

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (C) 2021 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okio.resourcefilesystem
+
+import okio.ExperimentalFileSystem
+import okio.FileMetadata
+import okio.FileNotFoundException
+import okio.FileSystem
+import okio.Path
+import okio.Path.Companion.toOkioPath
+import okio.Path.Companion.toPath
+import okio.Source
+import okio.internal.ReadOnlyFilesystem
+import okio.zipfilesystem.ZipFileSystem
+import java.io.File
+import java.util.concurrent.ConcurrentHashMap
+
+/**
+ * A file system exposing the traditional java classpath resources.
+ *
+ * Both metadata and file listings are best effort, and will work better
+ * for local project paths. The file system does not handle merging of
+ * multiple paths from difference resources like overlapping Jar files.
+ *
+ * ResourceFileSystem does not list classes, but will allow them to be read if specifically
+ * fetched.
+ */
+@ExperimentalFileSystem
+class ResourceFileSystem internal constructor(private val classLoader: ClassLoader) :
+  ReadOnlyFilesystem() {
+  var jarCache = ConcurrentHashMap<Path, ZipFileSystem>()
```

Comment:
Are we ok with caching this? It's probably strictly on top of classLoader.getResource(resourceName) loading and caching some resources.